macros/core/src/service.rs

```diff
@@ -1,413 +1,554 @@
+use parity_scale_codec::Encode;
+use proc_macro2::{Span, TokenStream};
+use proc_macro_error::abort;
+use quote::{quote, ToTokens};
+use std::collections::BTreeMap;
+use syn::{
+    parse::{Parse, ParseStream},
+    punctuated::Punctuated,
+    spanned::Spanned,
+    Expr, GenericArgument, Ident, ImplItemFn, ItemImpl, Path, PathArguments, Result as SynResult,
+    Token, Type, TypeParamBound, Visibility, WhereClause, WherePredicate,
+};
+
+static mut SERVICE_SPANS: BTreeMap<String, Span> = BTreeMap::new();
+
+pub fn gservice(args: TokenStream, service_impl: TokenStream) -> TokenStream {
+    let service_impl = parse_gservice_impl(service_impl);
+    ensure_single_gservice_on_impl(&service_impl);
+    ensure_single_gservice_by_name(&service_impl);
+    gen_gservice_impl(args, service_impl)
+}
+
+#[doc(hidden)]
+pub fn __gservice_internal(args: TokenStream, service_impl: TokenStream) -> TokenStream {
+    let service_impl = parse_gservice_impl(service_impl);
+    gen_gservice_impl(args, service_impl)
+}
+
+fn parse_gservice_impl(service_impl_tokens: TokenStream) -> ItemImpl {
+    syn::parse2(service_impl_tokens).unwrap_or_else(|err| {
+        abort!(
+            err.span(),
+            "`gservice` attribute can be applied to impls only: {}",
+            err
+        )
+    })
+}
+
+fn ensure_single_gservice_on_impl(service_impl: &ItemImpl) {
+    let attr_gservice = service_impl.attrs.iter().find(|attr| {
+        attr.meta
+            .path()
+            .segments
+            .last()
+            .map(|s| s.ident == "gservice")
+            .unwrap_or(false)
+    });
+    if attr_gservice.is_some() {
+        abort!(
+            service_impl,
+            "multiple `gservice` attributes on the same impl are not allowed",
+        )
+    }
+}
+
+fn ensure_single_gservice_by_name(service_impl: &ItemImpl) {
+    let path = shared::impl_type_path(service_impl);
+    let type_ident = path.path.segments.last().unwrap().ident.to_string();
+    if unsafe { SERVICE_SPANS.get(&type_ident) }.is_some() {
+        abort!(
+            service_impl,
+            "multiple `gservice` attributes on a type with the same name are not allowed"
+        )
+    }
+    unsafe { SERVICE_SPANS.insert(type_ident, service_impl.span()) };
+}
+
+fn gen_gservice_impl(args: TokenStream, service_impl: ItemImpl) -> TokenStream {
+    let service_args = syn::parse2::<ServiceArgs>(args).unwrap_or_else(|err| {
+        abort!(
+            err.span(),
+            "failed to parse `gservice` attribute arguments: {}",
+            err
+        )
+    });
+
+    let (service_type, service_type_args, service_type_constraints) = {
+        let service_type = ImplType::new(&service_impl);
+        (
+            service_type.path().clone(),
+            service_type.args().clone(),
+            service_type.constraints().cloned(),
+        )
+    };
+
+    let service_handlers = discover_service_handlers(&service_impl);
+
+    if service_handlers.is_empty() {

```

Comment:
Can service only expose other services without having  its own impl?

Reply:
Good question. I think I'll make it so a service should have its own methods or extend at least one service, or both

Reply:
Adjusted